bitcoin/core/__init__.py

```diff
@@ -446,6 +446,19 @@ def GetTxid(self):
             txid = Hash(self.serialize())
         return txid
 
+    def calc_weight(self):
+        """Calculate the transaction weight, as defined by BIP141.
+
+        The transaction must contain at least one input and one output.
+        """
+        assert len(self.vin) > 0
+        assert len(self.vout) > 0
+        if self.wit != CTxWitness():
```

Comment:
This check is not actually needed, because the transaction with empty witness will be the same as the 'stripped' transaction below. Also this condition will be false for mutable transaction without witness, because `self.wit` will be `CMutableTxWitness()`. There's also possibility that it will be `CTxWitness(CTxInWitness())`.

If we really need this check, it is better to use `not self.wit.is_null()` - it is used in `CTransaction.stream_serialize()` to check for empty witness.

Note that `GetTxId()` also checks with `self.wit != CTxWitness()` and I think it is also better to use `note self.wit.is_null()` there

Reply:
So for the empty witness case, I was thinking it'd be better not to serialize twice. But good point on the `is_null()`; fixed.